### DISCUSSION TITLE
FlorisBoard 0.4: "Auto-space after punctuation" discussion thread

### DISCUSSION THREAD

**Discussion by patrickgold** [florisboard/florisboard#1935] [General]
Beginning with 0.4.0-alpha01 FlorisBoard will have support for auto-spacing, at first only for Latin-based scripts. Auto-spacing after punctuation means that FlorisBoard will, for certain predefined symbols, automatically insert a space after the symbol. This should improve the typing experience, especially for glide typing and indirectly also for auto-capitalization, as it requires fewer manual space bar presses.

Auto-spacing is disabled by default for now (due to its potential of being quite annoying if it is not properly configured). If you want to try it out, go to Settings > Typing > Corrections > Auto-space after punctuation and enable it.

This thread is intended to stay open during the whole 0.4 development to gather feedback and to improve this specific feature. Please note that auto-spacing is completely independent from the word suggestions logic, thus please only post feedback and improvement ideas relevant to auto-spacing itself.

---

Original issues:
#375 
#1202 
#1240 

Probably also relevant:
#335 
#475 
#1013 
#1303

**Comment by whatevenisthat** [comment]
I don't know if this is relevant to this thread since it happens in 0.4.0-alpha01 but regardless of whether auto-space in the settings is enabled or not, but if I type a single letter and then glide type a word, a space will not be automatically inserted between the two words.

For example pressing **a** then glide typing **tree** results in **atree** rather than **a tree** (how it worked in 0.3.16).

**Reply by patrickgold** [reply]
~I mean technically this is the job of the phantom spacing logic to provide spaces between glided words, but it is very closely related to auto-spacing too.~

~It is interesting to me that you report this, as in the last few weeks I got several reports that glide typing does not perform auto-spacing correctly (but then it also failed back in 0.3.16). Basically it is #1937 though for you too, isn't it? If so, could you please comment in linked issue and also provide details about your device (Android version, which ROM, Device model and FlorisBoard version), so I can investigate further into this bug?~

See comment below

**Reply by patrickgold** [reply]
Nevermind, I think your issue is newly introduced and not related to #1937: if there's only one character in total auto-space for gliding will refuse to work, but as soon as two characters or more in total are present, it works. Is this also the case for you?

**Reply by whatevenisthat** [reply]
Yes that was the issue I noticed.

**Reply by patrickgold** [reply]
Ah ok perfect, can you please file this as a separate bug report so I can then properly reference it in the PR which fixes this bug?

**Comment by Glitchy-Tozier** [comment]
This was probably intended anyways, but I'd like to mention that this feature should be optional and off by default (as it manipulates/changes input).

**Reply by patrickgold** [reply]
Atm it is only temporarily off by default due to it being experimental. But now that I think about it, keeping this off by default forever is the best way to go, as it will always lead to unintended behavior for someone, no matter how often I improve this feature. Thanks for this reminder!

**Comment by creepy-pasta101** [comment]
Well..  typing an email address gives off an unintentional bug...... 
It becomes..  `username@domain. Com` automatically .....

**Reply by patrickgold** [reply]
That's true, if you type an email address in a normal text field this will happen. Will try to fix this in the next alpha release.

**Reply by Glitchy-Tozier** [reply]
That'll need to be a setting, right? I can't think of a way to differentiate URLs and email addresses from (weird) sentences that require an auto-space.

**Reply by patrickgold** [reply]
Not necessarily. It is true that I cannot differentiate between every possible form of email and URL, however the dot between the host and TLD in an email is easily distinguishble. Also this is only an issue in normal fields, in email or URL specific fields autospace0 is turned iff after punctuation.

**Reply by Bellalove47** [reply]
<>

**Comment by papjul** [comment]
I don't know how this improves the glide typing experience as on my side, Florisboard is already adding the necessary spaces when glide typing is used and "Automatic space after punctuation" is disabled.

I would suggest that if you have the option enabled and you hit the spacebar after a punctuation, this should not add another space after the automatic one, but keep it to one. If you really need a second space then you would need to hit the spacebar a second time (but I don't think this case would happen really often, when was the last time you typed two spaces after a punctuation?). This would avoid inevitable double spacing by mistake, and that way you could enable the option by default as it would not change our habits.

**Reply by patrickgold** [reply]
That is actually an awesome idea,  because I personally test-use the auto-space after punctuation too and constantly add a space after the punctuation.

**Comment by HikaruIchijyo** [comment]
Dunno if this is glitch or what ik this is due to missing word suggesstion but still here is the video chk text.

Adding space does this samething i mean if i remove 0 & put space 0 gets added again
 >
https://user-images.githubusercontent.com/38470944/184505931-4ca33f90-564d-4f74-a49d-d3d2ac6aaebb.mp4

**Reply by patrickgold** [reply]
Not really a bug but a demo of the autocomplete. If it annoys you please disable suggestions until the first release comes out with suggestions support.

**Reply by HikaruIchijyo** [reply]
> Not really a bug but a demo of the autocomplete. If it annoys you please disable suggestions until the first release comes out with suggestions support.

got it thanks

**Reply by paritosh9199** [reply]
Hey @patrickgold where are we on the word suggestion feature?

**Reply by patrickgold** [reply]
@paritosh9199 currently still in the prep phase, latest release v0.4.0-alpha03 implements the UI for the incognito mode (relevant for suggestions to know if they are allowed to learn or only statically suggest) and modernizes and reworks the Smartbar. If nothing unexpectedly gets thrown at me again I should now be finished with the preps and focus directly on the suggestions core.

**Reply by paritosh9199** [reply]
Awesome 👍👍

**Comment by JackEdwardAuty** [comment]
The auto-space feature seems to give a space after an @ symbol sometimes but not always which is almost always not the intended behaviour. It happened on a login window. 
"User@email. Com" 
Furthermore it still adds the unnecessary space after the domain as reported 3 years ago. 

Great work on the keyboard though!
